docs/architecture/adr-036-blockchain-reactor-refactor.md

```diff
@@ -0,0 +1,406 @@
+}
+```
+
+The `Controller` is initialized by providing initial height (`startHeight`) from which it will start downloading blocks from peers.
+
+``` go
+func ControllerInit(state ControllerState, startHeight int64) ControllerState {
+	state.Height = startHeight
+	state.State = FastSyncMode
+	state.MaxRequestPending = startHeight - 1
+	state.PendingRequestsNum = 0
+}
+
+func ControllerHandle(event Event, state ControllerState) (ControllerState, Message, TimeoutTrigger, Error) {
+	msg = nil
+	timeout = nil
+	error = nil
+
+	switch state.State {
+	case StateConsensusMode:
+		switch event := event.(type) {
+		case EventBlockRequest:
+			msg = createBlockResponseMessage(state, event)
+			return state, msg, timeout, error
+		default:
+			error = "Only respond to BlockRequests while in ConsensusMode!"
+			return state, msg, timeout, error
+		}
+
+	case StateFastSyncMode:
+		switch event := event.(type) {
+		case EventBlockRequest:
+			msg = createBlockResponseMessage(state, event)
+			return state, msg, timeout, error
+
+		case EventStatusReport:
+			if _, ok := state.PeerMap[event.PeerID]; !ok {
+				peerStats = PeerStats{-1, -1}
+			} else {
+				peerStats = state.PeerMap[event.PeerID]
+			}
+
+			if event.Height > peerStats.Height {
+				peerStats.Height = event.Height
+			}
+			if peerStats.PendingRequest == -1 {
+				msg = createBlockRequestMessage(state, event.PeerID, peerStats.Height)
+				if msg != nil {
+					peerStats.PendingRequest = msg.Height
+					state.PendingRequestsNum++
+					timeout = ResponseTimeoutTrigger{msg.PeerID, msg.Height, PeerTimeout}
+				} else if state.PendingRequestsNum == 0 {
+					timeout = TerminationTimeoutTrigger{state.Height, TerminationTimeout}
+				}
+			}
+			state.PeerMap[event.PeerID] = peerStats
+			return state, msg, timeout, error
+
+		case EventRemovePeer:
+			if _, ok := state.PeerMap[event.PeerID]; ok {
+				pendingRequest = state.PeerMap[event.PeerID].PendingRequest
+				if pendingRequest != -1 {
+					add(state.FailedRequests, pendingRequest)
+				}
+				state.PendingRequestsNum--
+				delete(state.PeerMap, event.PeerID)
+				if state.PeerMap.isEmpty() || state.PendingRequestsNum == 0 {
+					timeout = TerminationTimeoutTrigger{state.Height, TerminationTimeout}
+				}
+			} else {
+				error = "Removing unknown peer!"
+			}
+			return state, msg, timeout, error
+
+		case EventNewBlock:
+			if state.PeerMap[event.PeerID] {
+				peerStats = state.PeerMap[event.PeerID]
+				if peerStats.PendingRequest == event.Height {
+					peerStats.PendingRequest = -1
+					state.PendingRequestsNum--
+					if event.CurrentHeight > peerStats.Height {
+						peerStats.Height = event.CurrentHeight
+					}
+					state.Store[event.Height] = BlockInfo{event.Block, event.Commit, event.PeerID}
+					state = verifyBlocks(state) // it can lead to event.PeerID being removed from peer list
+
+					if _, ok := state.PeerMap[event.PeerID]; ok {
+						msg = createBlockRequestMessage(state, event.PeerID, peerStats.Height)
```

Comment:
Previous blockchain reactor will send multi block requests in the same time, but new version only send new request until last block have received, will it slow down?

Reply:
With the new version, there is one pending request per peer. After a response (block) arrives, we will try to send a new request to the peer. This makes protocol more reactive towards responsive peers as the peers that send responses faster will be asked more. Having multiple pending requests (especially big number of requests as in the current version) is problematic as it opens the space for DDoS attacks. It is also not clear that having many pending requests is indeed more efficient. We will need to measure this. The main goal of this ADR is correctness and improved testability, and this should hopefully make performance improvements easier after.